appholder/src/main/java/com/android/mdl/app/documentdata/DocumentDataReader.kt

```diff
@@ -0,0 +1,249 @@
+                val keys = (dataItem as Map).keys
+                if (keys.isEmpty()) {
+                    sb.append("{}")
+                } else {
+                    sb.append("{$newLine$indentString")
+                    for ((count, key) in keys.withIndex()) {
+                        sb.append("$space$space")
+                        val value = dataItem[key]
+                        cborPrettyPrintDataItem(sb, indent + 2, key)
+                        sb.append(" : ")
+                        cborPrettyPrintDataItem(sb, indent + 2, value)
+                        if (count + 1 < keys.size) {
+                            sb.append(",")
+                        }
+                        sb.append("$newLine $indentString")
+                    }
+                    sb.append("}")
+                }
+            }
+
+            MajorType.TAG -> throw java.lang.IllegalStateException("Semantic tag data item not expected")
+            MajorType.SPECIAL ->                 // Major type 7: floating point numbers and simple data types that need no
+                // content, as well as the "break" stop code.
+                if (dataItem is SimpleValue) {
+                    when (dataItem.simpleValueType) {
+                        SimpleValueType.FALSE -> sb.append("false")
+                        SimpleValueType.TRUE -> sb.append("true")
+                        SimpleValueType.NULL -> sb.append("null")
+                        SimpleValueType.UNDEFINED -> sb.append("undefined")
+                        SimpleValueType.RESERVED -> sb.append("reserved")
+                        SimpleValueType.UNALLOCATED -> sb.append("unallocated")
+                    }
+                } else if (dataItem is DoublePrecisionFloat) {
+                    val df = DecimalFormat(
+                        "0",
+                        DecimalFormatSymbols.getInstance(Locale.ENGLISH)
+                    )
+                    df.maximumFractionDigits = 340
+                    sb.append(df.format(dataItem.value))
+                } else if (dataItem is AbstractFloat) {
+                    val df = DecimalFormat(
+                        "0",
+                        DecimalFormatSymbols.getInstance(Locale.ENGLISH)
+                    )
+                    df.maximumFractionDigits = 340
+                    sb.append(df.format(dataItem.value))
+                } else {
+                    sb.append("break")
+                }
+        }
+    }
+
+    // Returns true iff all elements in |items| are not compound (e.g. an array or a map).
+    private fun cborAreAllDataItemsNonCompound(items: List<DataItem>): Boolean {
+        for (item in items) {
+            when (item.majorType) {
+                MajorType.ARRAY, MajorType.MAP -> return false
+                else -> {
+                }
+            }
+        }
+        return true
+    }
+
+    companion object {
+        private const val MDL_DOCTYPE = "org.iso.18013.5.1.mDL"
```

Comment:
Should we import these constants from DocumentData object?

Reply:
Sure!